### PR TITLE
Migrate spark table to iceberg table

### PR DESCRIPTION
**Pull request by chenjunjiedada**
This PR converts a spark table to iceberg table. It assumes no parallel operation is going on the original table and target table and thus is not thread-safe.

PR is still working in progress. Unit tests will be added when APIs is settled down.

### CODE REVIEW

**Comment by chenjunjiedada** [comment]
@rdblue @aokolnychyi, could you please help to take an early look? I 'd like to get some basic feedback firstly.

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala [hunk 1]
```diff
@@ -297,5 +301,88 @@ object SparkTableUtil {
...
+
+    val format = tableMetadata.provider.getOrElse("none")
+    if (format != "avro" && format != "parquet" && format != "orc") {
+      throw new UnsupportedOperationException(s"Unsupported format: $format")
+    }
+
+    val location = tableMetadata.location.toString
+    val partitionSpec = SparkSchemaUtil.specForTable(sparkSession, s"$dbName.$tableName")
+
+    val fastAppender = table.newFastAppend()
```
Comment (rdblue) [inline spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: I don't think there is a reason to use fast appends.

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala [hunk 2]
```diff
@@ -297,5 +301,88 @@ object SparkTableUtil {
...
+    val sparkSession = SparkSession.builder().getOrCreate()
+
+    if (!sparkSession.catalog.tableExists(dbName, tableName)) {
+      throw new NoSuchTableException(s"Table $dbName.$tableName does not exist")
+    }
+
+    val tableMetadata = sparkSession.sessionState.catalog.
+      getTableMetadata(new TableIdentifier(tableName, Some(dbName)))
+
+    val format = tableMetadata.provider.getOrElse("none")
```
Comment (rdblue) [inline spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: This should convert any Hive table, not just Spark DataSource tables. That's why the `listPartition` method supports a format for each partition. Use `partitionDF` to get a dataframe with each partition and format.

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala [hunk 3]
```diff
@@ -297,5 +301,88 @@ object SparkTableUtil {
...
+      throw new UnsupportedOperationException(s"Unsupported format: $format")
+    }
+
+    val location = tableMetadata.location.toString
+    val partitionSpec = SparkSchemaUtil.specForTable(sparkSession, s"$dbName.$tableName")
+
+    val fastAppender = table.newFastAppend()
+
+    val partitions = sparkSession.sessionState.catalog.externalCatalog
+      .listPartitionNames(dbName, tableName)
```
Comment (rdblue) [inline spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: This should use `partitionDF` and call `listPartition` in parallel for large tables. See the example notebook: https://github.com/apache/incubator-iceberg/blob/master/examples/Convert%20table%20to%20Iceberg.ipynb

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala [hunk 4]
```diff
@@ -297,5 +301,88 @@ object SparkTableUtil {
...
+        SparkTableUtil.listPartition(partitionToMap(e), location + "/" + e, format)
+      }
+
+      // Append manifest for each partition
+      dataFiles.foreach { partition =>
+        fastAppender.appendManifest(buildManifest(table, partition, partitionSpec))
+      }
+    }
+
+    fastAppender.apply()
```
Comment (rdblue) [inline spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: There's no need to call apply. Just commit.

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala [hunk 5]
```diff
@@ -297,5 +301,88 @@ object SparkTableUtil {
...
+
+  def partitionToMap(partition: String): Map[String, String] = {
+    val map = new mutable.HashMap[String, String]()
+    val list = partition.split("/")
+    list.foreach { str =>
+      val kv = str.split("=")
+      map.put(kv(0), kv(1))
+    }
+
+    map.toMap
```
Comment (jerryshao) [inline spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]:
```scala
partitions.split("/").map { s =>
  val kv = s.split("=")
  (kv(0), kv(1))
}.toMap
```

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala [hunk 6]
```diff
@@ -297,5 +301,88 @@ object SparkTableUtil {
...
+   * operation is going on original table and target table and thus is not
+   * thread-safe.
+   *
+   * @param dbName the database name of the table to be migrated
+   * @param tableName the table to be migrated
+   * @param table the target table to migrate in
+   *
+   * @return table the target table
+   */
+  def migrateSparkTable(dbName: String, tableName: String, table: Table): Table = {
```
Comment (aokolnychyi) [inline spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: Would it make sense to accept Spark `TableIdentifier` here instead of db and table separately?

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala [hunk 7]
```diff
@@ -297,5 +301,88 @@ object SparkTableUtil {
...
+    val sparkSession = SparkSession.builder().getOrCreate()
+
+    if (!sparkSession.catalog.tableExists(dbName, tableName)) {
+      throw new NoSuchTableException(s"Table $dbName.$tableName does not exist")
+    }
+
+    val tableMetadata = sparkSession.sessionState.catalog.
```
Comment (aokolnychyi) [inline spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: Does it make any difference for us whether the table is managed or external? What if we migrated to Iceberg and then the original table is dropped together with data?

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala (hunk 6 above)
Comment (aokolnychyi) [inline spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: The logic in this method sounds more like "import data" rather than "migrate" to me. Migration should probably replace a pointer in the catalog.

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala (hunk 3 above)
Comment (chenjunjiedada) [reply spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: I see.  One thing needs to confirm is do we need to use repartition to avoid creating too many manifests? Since in this case, we'd like to commit once after the spark job, so we don't need repartition.

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala [hunk 8]
```diff
@@ -297,5 +301,88 @@ object SparkTableUtil {
...
+
+    val location = tableMetadata.location.toString
+    val partitionSpec = SparkSchemaUtil.specForTable(sparkSession, s"$dbName.$tableName")
+
+    val fastAppender = table.newFastAppend()
+
+    val partitions = sparkSession.sessionState.catalog.externalCatalog
+      .listPartitionNames(dbName, tableName)
+    if (partitions.isEmpty) {
+      val dataFiles = SparkTableUtil.listPartition(Map.empty[String, String], location, format)
```
Comment (aokolnychyi) [inline spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: @rdblue Not related to this PR directly, but do we want to pass the Hadoop config from Spark to `listPartition` rather than calling `new Configuration()` inside `listPartition`?

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala (hunk 6 above)
Comment (chenjunjiedada) [reply spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: Make sense. I will update in next commit.

Comment (chenjunjiedada) [reply spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: Aha, it does like the "import data" behaviour. When I tried to use the iceberg, the first thing I had to do is to "import table".  If we want to implement a migration include replace the pointer, more steps should be needed. I can have a look further.  At this stage, I 'd like to implement "import data" since that should also be needed at first in the migration IIUC.

Comment (aokolnychyi) [reply spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: I totally agree. I think we cannot really implement "migrate" before we have proper DS V2 APIs coming in Spark 3.0. My comment was only about naming.

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala (hunk 7 above)
Comment (chenjunjiedada) [reply spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: It depends on whether the "replace pointer" action can happen on a managed or external table.  For a managed table, the table should be operated according to iceberg metadata after migration. For an external table, I 'm not sure whether we can replace the pointer, or whether we are allowed to replace the pointer.  In a real scenario, users may want to keep both of them, so we might provide an option to delete the pointer if we can.

**Comment by chenjunjiedada** [comment]
@rdblue, @aokolnychyi @jerryshao PR updated.

It uses `coalesce(1)` to append all files to one manifest file.  In the case of a large table which has a mass of files, the final manifest may be too big.  Currently, I'm not sure what is the best way to transfer manifests between the executor and the driver. IIUC, there are two ways if we want to build manifests parallel and commit once.

1.  Store the temporary manifests to some HDFS location and return location string to the driver, then the driver constructs the manifests and appends manifests and commits. While I don't find an existing API to build manifests from location, ManifestReader?  

2. Implement `Serializable ` for the Manifest or Snapshot, so that we collect them and append manifests in the driver side.  While this looks like a bit complex.

What do you think?

**Comment by chenjunjiedada** [comment]
Not sure why unit tests failed. 
`
org.apache.iceberg.spark.source.TestIcebergSourceHiveTables > testHiveManifestsTable FAILED
    java.lang.RuntimeException at TestIcebergSourceHiveTables.java:438
        Caused by: org.apache.thrift.transport.TTransportException at TestIcebergSourceHiveTables.java:438
            Caused by: java.net.SocketException at TestIcebergSourceHiveTables.java:438
org.apache.iceberg.spark.source.TestIcebergSourceHiveTables > testHiveHistoryTable FAILED
    java.lang.RuntimeException at TestIcebergSourceHiveTables.java:285
        Caused by: org.apache.thrift.transport.TTransportException at TestIcebergSourceHiveTables.java:285
            Caused by: java.net.SocketException at TestIcebergSourceHiveTables.java:285
org.apache.iceberg.spark.source.TestIcebergSourceHiveTables > testHiveEntriesTable FAILED
    java.lang.RuntimeException at TestIcebergSourceHiveTables.java:144
        Caused by: org.apache.thrift.transport.TTransportException at TestIcebergSourceHiveTables.java:144
            Caused by: java.net.SocketException at TestIcebergSourceHiveTables.java:144
org.apache.iceberg.spark.source.TestSparkTableUtil > testMigrateSparkTable FAILED
    org.apache.hadoop.hive.ql.metadata.HiveException at TestSparkTableUtil.java:134
        Caused by: org.apache.thrift.transport.TTransportException at TestSparkTableUtil.java:134
            Caused by: java.net.SocketException at TestSparkTableUtil.java:134
135 tests completed, 4 failed, 4 skipped
> Task :iceberg-spark:test FAILED
FAILURE: Build failed with an exception.
`

**Comment by chenjunjiedada** [comment]
hmm, it's a warning.
```
WARN org.apache.hadoop.hive.metastore.ObjectStore - Direct SQL failed, falling back to ORM
java.lang.ClassCastException: org.apache.derby.impl.jdbc.EmbedClob cannot be cast to java.lang.String
```

**Comment by chenjunjiedada** [comment]
Figured out the root cause of the failed unit test.  The spark job needs at least three thrift server threads since we have three partitions in the unit test, the maximum value of 5 should be enough while the previous unit tests still not release yet (default release time is 1800s).

File: hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java [hunk 9]
```diff
@@ -102,7 +102,7 @@ private TServer newThriftServer(TServerSocket socket, HiveConf conf) throws Exce
         .transportFactory(new TTransportFactory())
         .protocolFactory(new TBinaryProtocol.Factory())
         .minWorkerThreads(3)
-        .maxWorkerThreads(5);
+        .maxWorkerThreads(8);
```
Comment (rdblue) [inline hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java]: Why was this change required?

File: hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java [hunk 10]
```diff
@@ -111,6 +111,7 @@ private HiveConf newHiveConf(int port) {
     HiveConf newHiveConf = new HiveConf(new Configuration(), TestHiveMetastore.class);
     newHiveConf.set(HiveConf.ConfVars.METASTOREURIS.varname, "thrift://localhost:" + port);
     newHiveConf.set(HiveConf.ConfVars.METASTOREWAREHOUSE.varname, "file:" + hiveLocalDir.getAbsolutePath());
+    newHiveConf.set(HiveConf.ConfVars.METASTORE_TRY_DIRECT_SQL.varname, "false");
```
Comment (rdblue) [inline hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java]: What does this do?

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala [hunk 11]
```diff
@@ -19,18 +19,22 @@
...
+import com.google.common.collect.ImmutableMap
 import com.google.common.collect.Maps
 import java.nio.ByteBuffer
 import java.util
+import java.util.UUID
 import org.apache.hadoop.conf.Configuration
 import org.apache.hadoop.fs.{Path, PathFilter}
-import org.apache.iceberg.{DataFile, DataFiles, Metrics, MetricsConfig, PartitionSpec}
-import org.apache.iceberg.hadoop.HadoopInputFile
+import org.apache.iceberg._
```
Comment (rdblue) [inline spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: Please don't use wildcard imports.

File: hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java (hunk 9 above)
Comment (chenjunjiedada) [reply hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java]: This is used for the unit test. The spark job needs at least three thrift server threads since we have three partitions in the unit test, the maximum value of 5 should be enough while the previous unit tests still not release yet (default release time is 1800s) which cause failed to connect to thrift server.

File: hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java (hunk 10 above)
Comment (chenjunjiedada) [reply hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java]: This is used to eliminate warning in the unit test.

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala (hunk 11 above)
Comment (chenjunjiedada) [reply spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: hmm... The Idea IDE always optimizes the imports when adding a new one from Alt+Enter. Let me update it.

Comment (chenjunjiedada) [reply spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: It looks like we need to import 9 entities which exceed 100 line length limit. From spark's coding style, when importing more than 6 entities it prefers to use a wildcard. Just want to confirm which one do we prefer? line break or following spark coding style for scala?

Comment (rdblue) [reply spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: We never use wildcard imports.

File: hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java (hunk 10 above)
Comment (rdblue) [reply hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java]: What warning?

File: hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java (hunk 9 above)
Comment (rdblue) [reply hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java]: What have previous tests not released? What is the "default release time"?

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala [hunk 12]
```diff
@@ -297,5 +301,81 @@ object SparkTableUtil {
       )
     }
   }
+
+  private def buildManifest(table: Table,
+                    sparkDataFiles: Seq[SparkDataFile],
```
Comment (rdblue) [inline spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: Nit: indentation is off.

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala [hunk 13]
```diff
@@ -297,5 +301,81 @@ object SparkTableUtil {
...
+                    partitionSpec: PartitionSpec): ManifestFile = {
+    val outputFile = table.io
+      .newOutputFile(FileFormat.AVRO.addExtension("/tmp/" + UUID.randomUUID.toString))
+    val writer = ManifestWriter.write(partitionSpec, outputFile)
+    try {
+      for (file <- sparkDataFiles) {
```
Comment (rdblue) [inline spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: Style: please use `foreach` instead of `for` loops in scala.

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala [hunk 14]
```diff
@@ -297,5 +301,81 @@ object SparkTableUtil {
...
+    }
+
+    val partitionSpec = SparkSchemaUtil.specForTable(sparkSession, s"$dbName.$tableName")
+    val conf = sparkSession.sparkContext.hadoopConfiguration
+    val tables = new HadoopTables(conf)
+    val schema = SparkSchemaUtil.schemaForTable(sparkSession, s"$dbName.$tableName")
+    val table = tables.create(schema, partitionSpec, ImmutableMap.of(), location)
+    val appender = table.newAppend()
+
+    if (partitionSpec == PartitionSpec.unpartitioned) {
```
Comment (rdblue) [inline spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: Why is there a special case for unpartitioned tables? Doesn't `partitionDF` return a single partition in that case?

File: hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java (hunk 9 above)
Comment (chenjunjiedada) [reply hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java]: It is a new unit test I added, not related to the previous.

The default release time is referred to `hive.server2.thrift.http.worker.keepalive.time` which is the `stopTimeoutVal` in TThreadPoolServer and is set default 60s ( I referred to `hive.server2.thrift.http.max.idle.time` by mistake..).

File: hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java (hunk 10 above)
Comment (chenjunjiedada) [reply hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java]: WARN org.apache.hadoop.hive.metastore.ObjectStore - Direct SQL failed, falling back to ORM
java.lang.ClassCastException: org.apache.derby.impl.jdbc.EmbedClob cannot be cast to java.lang.String.

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala (hunk 11 above)
Comment (chenjunjiedada) [reply spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: I see, will update in next commit.

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala (hunk 14 above)
Comment (chenjunjiedada) [reply spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: Yes, here is the exception:
```
org.apache.hadoop.hive.ql.metadata.HiveException: Table item is not a partitioned table
  at org.apache.hadoop.hive.ql.metadata.Hive.getPartitions(Hive.java:2123)
  at org.apache.hadoop.hive.ql.metadata.Hive.getPartitions(Hive.java:2156)
  at org.apache.spark.sql.hive.client.HiveClientImpl$$anonfun$getPartitions$1.apply(HiveClientImpl.scala:670)
  at org.apache.spark.sql.hive.client.HiveClientImpl$$anonfun$getPartitions$1.apply(HiveClientImpl.scala:662)
  at org.apache.spark.sql.hive.client.HiveClientImpl$$anonfun$withHiveState$1.apply(HiveClientImpl.scala:275)
  at org.apache.spark.sql.hive.client.HiveClientImpl.liftedTree1$1(HiveClientImpl.scala:213)
  at org.apache.spark.sql.hive.client.HiveClientImpl.retryLocked(HiveClientImpl.scala:212)
  at org.apache.spark.sql.hive.client.HiveClientImpl.withHiveState(HiveClientImpl.scala:258)
  at org.apache.spark.sql.hive.client.HiveClientImpl.getPartitions(HiveClientImpl.scala:662)
  at org.apache.spark.sql.hive.client.HiveClient$class.getPartitions(HiveClient.scala:210)
  at org.apache.spark.sql.hive.client.HiveClientImpl.getPartitions(HiveClientImpl.scala:83)
  at org.apache.iceberg.spark.hacks.Hive.partitions(Hive.java:48)
  at org.apache.iceberg.spark.SparkTableUtil$.partitionDF(SparkTableUtil.scala:56)
```

If that is not expected, I can try to fix that first

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala (hunk 13 above)
Comment (chenjunjiedada) [reply spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: OK, will update.

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala [hunk 15]
```diff
@@ -297,5 +302,81 @@ object SparkTableUtil {
       )
     }
   }
+
+  private def buildManifest(table: Table,
+      sparkDataFiles: Seq[SparkDataFile],
+      partitionSpec: PartitionSpec): ManifestFile = {
+    val outputFile = table.io
+      .newOutputFile(FileFormat.AVRO.addExtension("/tmp/" + UUID.randomUUID.toString))
```
[the remark this answers is not in the document]
Comment (chenjunjiedada) [reply spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: Agreed. Just want to confirm why it can't be the local file? It is just a temporary file, I wanted to store it in the tmpfs of localhost to save some IO.

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala [hunk 16]
```diff
@@ -297,5 +302,81 @@ object SparkTableUtil {
...
+    val tables = new HadoopTables(conf)
+    val schema = SparkSchemaUtil.schemaForTable(sparkSession, s"$dbName.$tableName")
+    val table = tables.create(schema, partitionSpec, ImmutableMap.of(), location)
+    val appender = table.newAppend()
+
+    if (partitionSpec == PartitionSpec.unpartitioned) {
+      val tableMetadata = sparkSession.sessionState.catalog.getTableMetadata(source)
+      val format = tableMetadata.provider.getOrElse("none")
+
+      if (format != "avro" && format != "parquet" && format != "orc") {
```
[the remark this answers is not in the document]
Comment (chenjunjiedada) [reply spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: ok, will update.

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala [hunk 17]
```diff
@@ -297,5 +302,81 @@ object SparkTableUtil {
...
+        throw new UnsupportedOperationException(s"Unsupported format: $format")
+      }
+      listPartition(Map.empty[String, String], tableMetadata.location.toString,
+        format).foreach{f => appender.appendFile(f.toDataFile(PartitionSpec.unpartitioned))}
+      appender.commit()
+    } else {
+      val partitions = partitionDF(sparkSession, s"$dbName.$tableName")
+      partitions.flatMap { row =>
+        listPartition(row.getMap[String, String](0).toMap, row.getString(1), row.getString(2))
+      }.coalesce(1).mapPartitions {
```
[the remark this answers is not in the document]
Comment (chenjunjiedada) [reply spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: That's a temporary use. I left comments to discuss a way to serialize manifest info [here](https://github.com/apache/incubator-iceberg/pull/374#issuecomment-522904779). Looks like the case class is a good way to do that, I will copy this.

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala (hunk 15 above)
[the remark this answers is not in the document]
Comment (chenjunjiedada) [reply spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: Understood, I was still thinking serialize the manifest file...

File: spark/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java [hunk 18]
```diff
@@ -116,4 +123,28 @@ public void testPartitionScanByFilter() {
     Dataset<Row> partitionDF = SparkTableUtil.partitionDFByFilter(spark, qualifiedTableName, "data = 'a'");
     Assert.assertEquals("There should be 1 matching partition", 1, partitionDF.count());
   }
+
+  @Test
+  public void testImportPartitionedTable() throws Exception {
```
[the remark this answers is not in the document]
Comment (chenjunjiedada) [reply spark/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java]: Currently, it creates a hadoop table inside. I can change the signature to accept `Table` parameter so that we can import as the specific table.  you mean test import as a hadoop table and import as a hive table, right?

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala [hunk 19]
```diff
@@ -131,18 +136,22 @@ object SparkTableUtil {
...
+          arrayToMap(columnSizes),
+          arrayToMap(valueCounts),
+          arrayToMap(nullValueCounts),
+          arrayToMap(lowerBounds),
+          arrayToMap(upperBounds)))
+
+      if (partitionKey == "") {
+        builder.build()
+      } else {
+        builder.withPartitionPath(partitionKey).build()
```
Comment (aokolnychyi) [inline spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: Not directly related to PR: it seems we don't support dates in `withPartitionPath` as `fromPartitionString` in `Conversions` doesn't handle dates. @chenjunjiedada @rdblue am I correct?

Comment (aokolnychyi) [reply spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: Should we have something like below?

```
case DATE:
  return Literal.of(asString).to(Types.DateType.get()).value();
```

**Comment by chenjunjiedada** [comment]
Interesting, the unit test works when running in intellij, but failed here.

**Comment by chenjunjiedada** [comment]
close and open to trigger ci

**Comment by chenjunjiedada** [comment]
This time is hive test failed.
```
> Task :iceberg-hive:test
org.apache.iceberg.hive.HiveTableTest > testConcurrentConnections FAILED
    java.lang.AssertionError at HiveTableTest.java:330
```

File: hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java [hunk 20]
```diff
@@ -102,7 +102,7 @@ private TServer newThriftServer(TServerSocket socket, HiveConf conf) throws Exce
         .transportFactory(new TTransportFactory())
         .protocolFactory(new TBinaryProtocol.Factory())
         .minWorkerThreads(3)
-        .maxWorkerThreads(5);
+        .maxWorkerThreads(10);
```
Comment (aokolnychyi) [inline hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java]: Would it make sense to expose this as a config option in `TestHiveMetastore`? A small number of worker threads helped us to catch a few connection leaks early on.

File: spark/src/main/java/org/apache/iceberg/spark/hacks/Hive.java [hunk 21]
```diff
@@ -62,4 +63,16 @@ private Hive() {}
 
     return client.getPartitionsByFilter(client.getTable(db, table), exprs);
   }
+
+  public static CatalogTable getTable(SparkSession spark, String name) {
+    List<String> parts = Lists.newArrayList(Splitter.on('.').limit(2).split(name));
+    String db = parts.size() == 1 ? "default" : parts.get(0);
+    String table = parts.get(parts.size() == 1 ? 0 : 1);
+
+    HiveClient client = HiveUtils$.MODULE$.newClientForMetadata(
```
Comment (aokolnychyi) [inline spark/src/main/java/org/apache/iceberg/spark/hacks/Hive.java]: Is there a specific reason to use `HiveClient` directly? Would it make sense to use `spark.sessionState.catalog`? That way, we will also support the in-memory catalog and any other catalog that can return `CatalogTable` for us.

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala (hunk 19 above)
Comment (aokolnychyi) [reply spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: Let me create an issue for this. We can close it later on if we don't want to fix it.

File: spark/src/main/java/org/apache/iceberg/spark/hacks/Hive.java (hunk 21 above)
Comment (jerryshao) [reply spark/src/main/java/org/apache/iceberg/spark/hacks/Hive.java]: Agree to use Spark's session catalog rather than directly manipulating hive or other catalog implementations.

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala [hunk 22]
```diff
@@ -131,18 +136,22 @@ object SparkTableUtil {
...
+        .withFormat(format)
+        .withFileSizeInBytes(fileSize)
+        .withMetrics(new Metrics(rowCount,
+          arrayToMap(columnSizes),
+          arrayToMap(valueCounts),
+          arrayToMap(nullValueCounts),
+          arrayToMap(lowerBounds),
+          arrayToMap(upperBounds)))
+
+      if (partitionKey == "") {
```
Comment (jerryshao) [inline spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: nit: `partitionKey.isEmpty()`

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala [hunk 23]
```diff
@@ -297,5 +306,104 @@ object SparkTableUtil {
...
     }
   }
+
+  private def buildManifest(conf: SerializableConfiguration,
+      sparkDataFiles: Seq[SparkDataFile],
+      partitionSpec: PartitionSpec,
+      basePath: String): Iterator[Manifest] = {
+    if (sparkDataFiles.isEmpty) {
+      Seq.empty.iterator
+    }
```
Comment (jerryshao) [inline spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: This check will not return empty iterator directly. If you want a return here you should change to `return Seq.empty.iterator` (this is not scalastyle).

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala [hunk 24]
```diff
@@ -297,5 +306,104 @@ object SparkTableUtil {
...
+
+    val io = new HadoopFileIO(conf.get())
+    val ctx = TaskContext.get()
+    val location = new Path(basePath,
+      s"stage-${ctx.stageId()}-task-${ctx.taskAttemptId()}-manifest")
+    val outputFile = io.newOutputFile(FileFormat.AVRO.addExtension(location.toString))
+    val writer = ManifestWriter.write(partitionSpec, outputFile)
+    try {
+      sparkDataFiles.foreach { file =>
+        writer.add(file.toDataFile(partitionSpec))
```
Comment (jerryshao) [inline spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: What will be happened if an exception is raised here?

File: hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java (hunk 20 above)
Comment (chenjunjiedada) [reply hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java]: Agree to make it configurable. 

The small threads number is OK for me as long as it is enough for new tests. So I'd like to know if we can release the thread/connection when each unit test finish?

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala (hunk 22 above)
Comment (chenjunjiedada) [reply spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: updated.

File: spark/src/main/java/org/apache/iceberg/spark/hacks/Hive.java (hunk 21 above)
Comment (chenjunjiedada) [reply spark/src/main/java/org/apache/iceberg/spark/hacks/Hive.java]: OK, updated.

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala (hunk 24 above)
Comment (chenjunjiedada) [reply spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: The expected behaviour is spark retry the task. If one of the tasks fail eventually, the job is failed and manifests are discarded from the driver as well.  As a result, the partial result will not be committed.

**Comment by chenjunjiedada** [comment]
The unit tests passed. @rdblue you can continue to review.

File: hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java (hunk 20 above)
Comment (rdblue) [reply hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java]: I think that using Spark's catalog will avoid some of these issues because tests will reuse the same connection.

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala (hunk 23 above)
Comment (rdblue) [reply spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: Good catch. I assumed this method had not been changed much. This should be fixed by wrapping the write in an else case.

File: hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java (hunk 20 above)
Comment (chenjunjiedada) [reply hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java]: Make sense and the unit test passed when I change back to use Spark's catalog.

**Comment by chenjunjiedada** [comment]
The CI failed due to python unit tests not related to java.

File: spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala [hunk 25]
```diff
@@ -297,5 +305,104 @@ object SparkTableUtil {
       )
     }
   }
+
+  private def buildManifest(conf: SerializableConfiguration,
+      sparkDataFiles: Seq[SparkDataFile],
+      partitionSpec: PartitionSpec,
+      basePath: String): Iterator[Manifest] = {
+    if (sparkDataFiles.isEmpty) {
+      return Seq.empty.iterator
```
Comment (jerryshao) [inline spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: you don't need a `return` now for your current changes.

Comment (chenjunjiedada) [reply spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala]: I see... Thanks

**Comment by chenjunjiedada** [comment]
@rdblue, could you please help to have a look again? It is ready now.

File: hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java (hunk 20 above)
Comment (rdblue) [reply hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java]: Can you revert this change? I'd like to see if it works with the recent changes to reduce the number of clients in tests.

Comment (chenjunjiedada) [reply hive/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java]: Done.

**Comment by rdblue** [comment]
+1

Thanks, @chenjunjiedada! I'll merge this soon. I'm trying to test a fix for the flaky tests and I'd like to resume merging afterwards to avoid needing to restart so many tests.
